google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/AbstractSubscriberConnection.java

```diff
@@ -53,30 +44,30 @@
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
-import javax.annotation.Nullable;
 import org.joda.time.Duration;
 import org.joda.time.Instant;
 import org.joda.time.Interval;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-/** Implementation of {@link Subscriber}. */
-final class SubscriberConnection extends AbstractService {
-  private static final Logger logger = LoggerFactory.getLogger(SubscriberConnection.class);
+/**
+ * Abstract base implementation class of a subscriber connection in charge of receiving subscription
+ * messages.
+ */
+abstract class AbstractSubscriberConnection extends AbstractService {
```

Comment:
I so much agree with the principle where most of the uses of inheritance can be more safely converted to a containment model with all the advantages cited in Item 16 of the "effective java" book. But in this case I argue that the AbstractSubscriberConnection was designed specifically to be extended as noted as one valid uses of inheritance. Said that I switched to containment, take a look a let me know if it makes the code flow more understandable.